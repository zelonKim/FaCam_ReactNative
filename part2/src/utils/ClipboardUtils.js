import { getStringAsync } from "expo-clipboard"

export const getClipboardString = async () => {
    const result = await getStringAsync();
    return result;
}