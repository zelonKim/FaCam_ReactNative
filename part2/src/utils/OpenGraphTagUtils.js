import {OpenGraphParser} from 'react-native-opengraph-kit';

export const getOpenGraphData = async(url) => {
    const result = await OpenGraphParser.extractMeta(url) // OpenGraphParser.extractMeta(url): 해당 url 페이지에 대한 정보를 가져옴.
    console.log(result)
    /*
     [{"Referrer": "origin", 
        "title": "네이버", 
        "description": "네이버 메인에서 다양한 정보와 유용한 컨텐츠를 만나 보세요",
        "image": "https://s.pstatic.net/static/www/mobile/edit/2016/0705/mobile_212852414260.png", 
        "url": "https://www.naver.com",
        "viewport": "width=1190"
     }] 
    */

    return result[0] || null
}

