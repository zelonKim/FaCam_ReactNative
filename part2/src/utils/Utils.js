
export const getRandomSixNumber = () => {
    const numberArray = [];

    while(numberArray.length < 6) {
        const number = (Math.floor(Math.random() * 45)) + 1;
        const hasNumber = numberArray.filter((prev) => prev === number).length > 0;
    
        if(!hasNumber) numberArray.push(number);
    }

    return numberArray.sort((numA, numB) => numA - numB); // 오름차순 나열
}