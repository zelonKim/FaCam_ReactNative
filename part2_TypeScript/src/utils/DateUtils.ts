export const getMillisToDateString = (millis: number) => {
    const date = new Date(millis);

    return `${date.getFullYear()}-${(date.getMonth() + 1) < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1 }-${date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()}`
}