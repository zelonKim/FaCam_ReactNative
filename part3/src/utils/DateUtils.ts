export const convertToDateString = (time:number):string => {

    const date = new Date(time);
    
    const addZeroIfOneCharacter = (num:number):string => {
        if(num<10) {
            return `0${num}`
        }
        return num.toString();
    }

    return `${date.getFullYear()}-${addZeroIfOneCharacter(date.getMonth()+1)}-${addZeroIfOneCharacter(date.getDate())} / ${addZeroIfOneCharacter(date.getHours())}:${addZeroIfOneCharacter(date.getMinutes())}`
}