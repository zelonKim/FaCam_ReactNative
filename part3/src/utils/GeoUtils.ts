import { KAKAO_API_KEY } from "../../api_key";


export const getAddressFromCoords = (
    latitude: number,
    longitude: number
): Promise<string | null> => {
    return fetch(`https://dapi.kakao.com/v2/local/geo/coord2address.json?x=${longitude}&y=${latitude}`,
        {
            method: 'GET',
            headers: {
                Authorization: `KakaoAK ${KAKAO_API_KEY}`
            }
        }
    ).then(result => result.json())
     .then(result => {
        if(result.meta.total_count === 0) {
            return null;
        }
        if(result.documents.length === 0) {
            return null;
        }
        const addressItem = result.documents[0];

        return addressItem.address.address_name;
     })
}



export const getCoordsFromAddress = (address:string):Promise<{latitude:number, longitude:number, address:string} | null> => {
    return fetch(`https://dapi.kakao.com/v2/local/search/address.json?query=${address}`,
        {
            method: 'GET',
            headers: {
                Authorization: `KakaoAK ${KAKAO_API_KEY}`
            }
        }
    ).then(result => result.json())
     .then(result => {
        if(result.meta.total_count === 0) {
            return null;
        }
        if(result.documents.length === 0) {
            return null;
        }
        const addressItem = result.documents[0];

        return {
            latitude: addressItem.y,
            longitude: addressItem.x,
            address: addressItem.address_name
        } 
     })
}




export const getCoordsFromKeyword = (keyword:string):Promise<{latitude:number, longitude:number, address:string} | null> => {
    return fetch(`https://dapi.kakao.com/v2/local/search/keyword.json?query=${keyword}`,
        {
            method: 'GET',
            headers: {
                Authorization: `KakaoAK ${KAKAO_API_KEY}`
            }
        }
    ).then(result => result.json())
     .then(result => {
        if(result.meta.total_count === 0) {
            return null;
        }
        if(result.documents.length === 0) {
            return null;
        }
        const addressItem = result.documents[0];

        return {
            latitude: addressItem.y,
            longitude: addressItem.x,
            address: addressItem.address_name
        } 
     })
}



