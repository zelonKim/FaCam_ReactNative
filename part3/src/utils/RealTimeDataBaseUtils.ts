import database from '@react-native-firebase/database'

export const saveNewRestaurant = async(params:{
    title: string, 
    address: string,
    latitude: number,
    longitude: number,
}) => {
    const db = database().ref('/restaurant')
    const saveItem = {
        title: params.title,
        address: params.address,
        latitude: params.latitude,
        longitude: params.longitude
    }
    await db.push().set({...saveItem}) 
}; 


export const getRestaurantList = async(): Promise<{title: string, address: string, latitude:number, longitude: number}[]> => {
    const db = database().ref('/restaurant')
    const snapshotValue = await db.once('value').then((snapshot) => snapshot.val());

    return Object.keys(snapshotValue).map((key) => snapshotValue[key])
}