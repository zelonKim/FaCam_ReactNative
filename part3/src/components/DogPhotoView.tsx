import React, { useCallback, useEffect, useState } from "react";
import { TypeDog } from "../data/TypeDog";
import { RemoteImage } from "./RemoteImage";
import { useWindowDimensions } from "react-native";
import { api } from "../utils/AxiosUtils";


export const DogPhotoView: React.FC = () => {
    const [dog, setDog] = useState<TypeDog | null>(null);
    const {width} = useWindowDimensions();

    const loadDogPhoto = useCallback(async() => {
        const apiResult = await api.get<{message: string; status: string;}>('breeds/image/random');
        const result = apiResult.data;
        setDog({photoUrl: result.message});
    }, [])


    useEffect(() => {
        loadDogPhoto()
    }, [loadDogPhoto])


    if(dog === null) {
        return null;
    }

    return (
        <RemoteImage 
            testID="image" 
            url={dog.photoUrl}  
            width={width * 0.7} 
            height={width * 0.7} 
        />
    )
}