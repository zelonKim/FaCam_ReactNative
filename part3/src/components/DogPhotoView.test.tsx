import {render, waitFor} from '@testing-library/react-native';
import { DogPhotoView } from './DogPhotoView';
import MockAdapter from 'axios-mock-adapter';
import { api } from '../utils/AxiosUtils';


describe('DogPhoto 렌더링 테스트', () => {
    let mock: MockAdapter;

    beforeAll(() => {
        mock = new MockAdapter(api);
    })

    afterAll(() => {
        mock.reset();
    })


    test('DogPhotoView 렌더링 스냅샷', async() => {
        mock.onGet(new RegExp('/breeds/image/random')).reply(200, {
            message: 'TEST_MESSAGE',
            status: 'SUCCESS'
        })

        const component = render(<DogPhotoView />);

        await waitFor(() => component.findByTestId('image'), {timeout: 2000})

        expect(component.toJSON()).toMatchSnapshot();

    })

})
