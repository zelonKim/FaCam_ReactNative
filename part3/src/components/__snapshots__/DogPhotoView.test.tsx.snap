// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DogPhoto 렌더링 테스트 DogPhotoView 렌더링 스냅샷 1`] = `
<Image
  source={
    {
      "uri": "TEST_MESSAGE",
    }
  }
  style={
    [
      undefined,
      {
        "height": 525,
        "width": 525,
      },
    ]
  }
  testID="image"
/>
`;

exports[`DogPhotoView 렌더링 테스트 1`] = `
<Image
  source={
    {
      "uri": "https://images.dog.ceo/breeds/schipperke/n02104365_5670.jpg",
    }
  }
  style={
    [
      undefined,
      {
        "height": 525,
        "width": 525,
      },
    ]
  }
  testID="image"
/>
`;
